aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu

```diff
@@ -0,0 +1,326 @@
+template <typename scalar_t>
+__global__ void sum_and_scatter(
+    int64_t *input, scalar_t *gradWeight, int64_t stride,
+    int64_t* segment_offsets, int64_t num_of_segments, const scalar_t *grad_weight_per_segment,
+    const int64_t *segment_sizes_offsets, int64_t num_of_partial_segments,
+    const int64_t stride_warped) {
+
+  const int gid = blockIdx.x * blockDim.x + threadIdx.x;
+  const int id = gid / stride_warped;
+  const int startFeature = gid % stride_warped;
+  if (startFeature >= stride) {
+    return;
+  }
+  if (id >= num_of_segments) {
+    return;
+  }
+
+  const int idx_begin = segment_sizes_offsets[id];
+  const int idx_end = (id == num_of_segments-1)?num_of_partial_segments:segment_sizes_offsets[id+1];
+  acc_type<scalar_t, true> weight = 0;
+  for (int idx=idx_begin; idx < idx_end; ++idx) {
+    weight += grad_weight_per_segment[idx*stride + startFeature];
+  }
+  const int weightRow = input[segment_offsets[id]] * stride;
+  gradWeight[weightRow + startFeature] = weight;
+}
+
+} // anon namespace
+
+Tensor embedding_backward_cuda_kernel(
+        const Tensor &grad,
+        const Tensor &orig_indices,
+        const Tensor &sorted_indices,
+        const Tensor &count,
+        int64_t num_weights,
+        int padding_idx,
+        bool scale_grad_by_freq,
+        bool mode_mean,
+        const Tensor &offset2bag,
+        const Tensor &bag_size,
+        const Tensor &per_sample_weights) {
+
+  auto stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+  const ptrdiff_t numel = sorted_indices.numel();
+
+  auto grad_weight = at::zeros({num_weights, grad.size(-1)}, grad.options());
+  const int64_t stride = grad_weight.stride(0);
+
+  // Compute the number of segments and their start position so that we do not have to
+  // spawn a warp per index. In this context, a segment is a number of rows that should
+  // be summarized.
+  // Unit: index in `sorted_indices` and `orig_indices`
+  thrust::device_vector<int64_t> segment_offsets(numel);
+  int64_t num_of_segments;
+  {
+    auto sorted_indices_dev = thrust::device_ptr<int64_t>(sorted_indices.data<int64_t>());
+    auto dummy = at::empty_like(sorted_indices);
+    auto dummy_dev = thrust::device_ptr<int64_t>(dummy.data<int64_t>());
+    auto ends = thrust::unique_by_key_copy(
+            policy,
+            sorted_indices_dev,
+            sorted_indices_dev + numel,
+            thrust::make_counting_iterator(0),
+            dummy_dev,
+            thrust::raw_pointer_cast(segment_offsets.data()));
+    num_of_segments = thrust::get<0>(ends) - dummy_dev;
+  }
+
+  // We split the segments up into sizes of `NROWS_PER_THREAD`
+  // Compute the number partial-segments per segment (some partial-segments 
+  // may not be the full `NROWS_PER_THREAD` number of rows)
+  thrust::device_vector<int64_t> partials_per_segment(num_of_segments);
```

Comment:
thrust::device_vector with default allocator should never be used. It calls cudaMalloc on construction and cudaFree on destruction

Reply:
Fixed in #22401